main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -920,6 +988,51 @@ object Resolver extends Phase[NamedAst.Root, ResolvedAst.Root] {
     }
   }
 
+  /**
+    * Finds the signature with the qualified name `qname` in the namespace `ns0`.
+    */
+  def lookupSig(qname: Name.QName, ns0: Name.NName, prog0: NamedAst.Root): Validation[NamedAst.Sig, ResolutionError] = {
+    val sigOpt = tryLookupSig(qname, ns0, prog0)
+
+    sigOpt match {
+      case None => ResolutionError.UndefinedName(qname, ns0, qname.loc).toFailure
+      case Some(s) => getSigIfAccessible(s, ns0, qname.loc)
+    }
+  }
+
+  /**
+    * Tries to find a sig with the qualified name `qname` in the namespace `ns0`.
+    */
+  def tryLookupSig(qname: Name.QName, ns0: Name.NName, prog0: NamedAst.Root): Option[NamedAst.Sig] = {
+    // Check whether the name is fully-qualified.
+    if (qname.isUnqualified) {
+      // Case 1: Unqualified name. Lookup in the current namespace.
+      val sigOpt = findSig(prog0.classes.getOrElse(ns0, Map.empty), qname)
+
+      sigOpt match {
+        case Some(sig) =>
+          // Case 1.2: Found in the current namespace.
+          Some(sig)
+        case None =>
+          // Case 1.1: Try the global namespace.
+          findSig(prog0.classes.getOrElse(Name.RootNS, Map.empty), qname)
+      }
+    } else {
+      // Case 2: Qualified. Lookup in the given namespace.
+      findSig(prog0.classes.getOrElse(qname.namespace, Map.empty), qname)
+    }
+  }
+
+  /**
+    * Finds the signature with the given `qname` from among the given `classes`.
+    */
+  def findSig(classes: Map[String, NamedAst.Class], qname: Name.QName): Option[NamedAst.Sig] = {
```

Comment:
Iterating through classes like this is going to be too expensive. (Almost certainly this will kill performance).

I think you need to build a helper map `QName -> Class` or the like and carry that around.